### PR TITLE
Update contributing guide with etiquette section

### PR DESCRIPTION
**Pull request by michaelosthege**
Lately we have seen several examples where people asked for permission or commented that they wanted to work on an issue, but did not follow up with a pull request within several __weeks__ or at all.

Because this is blocking other people from getting the work done, and thereby slowing the overall progress, I'm proposing to add a corresponding section to our `CONTRIBUTING.md`.

The gist of it is:
+ We value _getting work done_ over pledges.
+ We ask people to open `Draft` PRs as soon as they start working on it.
+ We reserve the right to take over or close PRs when authors go unresponsive.

I think the risk of doing work twice is really low.

Also among high-frequency contributors pledges are rare - if happening at all they are usually more of the kind "I can do this after these two other things are merged".

And in the rare event of two people picking up the same issue at the same time (within 1-3 days) they can help out each other and decide which PR to go for.

-----------

Why are the tests triggering? I thought they only run if there are changes in `.py` files?
(I cancelled them to save resources.)

### CODE REVIEW

File: CONTRIBUTING.md
Comment (OriolAbril) [inline CONTRIBUTING.md:38]: first step should now be reading the etiquette section

File: CONTRIBUTING.md
Comment (OriolAbril) [inline CONTRIBUTING.md]:
```suggestion
Make sure to mention the issue number in the PR description.
```

I think here keeping it simple and not providing options the readers potentially won't know how to choose between is better.

File: CONTRIBUTING.md
Comment (OriolAbril) [inline CONTRIBUTING.md]: I think it would also be good to add something along the lines of work on open issues, don't submit PRs for new features that have not been discussed in an issue/discussion/discourse with the core contributors.

File: CONTRIBUTING.md
Comment (ricardoV94) [inline CONTRIBUTING.md]: They won't exactly be ignored... I guess what you want to convey is that an issue is not exclusive to someone else, just because they commented on it first. I think it is still polite to ask the original person if they are still planning to work on it or if they can take over. I've seen several people saying "I don't have time, thanks for asking. Yes take over"

Comment (ricardoV94) [reply CONTRIBUTING.md]: Exactly as you say, recurring contributors don't have the need to do this, because they know they will open a PR fast enough to not be worth it. But I think asking for "permission" is still something new contributors feel the need for, and I don't think it's necessarily bad.

Edit: I wouldn't want to foster an environment where people feel the need to "rush" to claim something. Also opening draft PRs early on wastes CI compute

**Comment by ricardoV94** [comment]
Also, can we remove the guideline to prefix a draft PR with [WIP] (mentioned a couple of lines below your changes)? We should just use the draft tool from Github, it's more clean.

File: CONTRIBUTING.md
Comment (OriolAbril) [inline CONTRIBUTING.md]: This paragraph feels a bit harsh and I think is prone to be misunderstood. Some suggestions:

Issues are not assigned nor reserved via issue comments. Our way of avoiding duplicated work is with draft PRs and
by not opening PRs that overlap with other _active PRs_. We will still track and be active answering any questions posted
on issues to help you open a PR for it.

_Active PRs_ means currently open PRs whose author has commented on or commited in the last week. Once a PR becomes inactive, core contributors can choose to take over or close the PR. The decision will depend on each case and the availability of core contributors and will not necessarly happen exactly on the 7th day of inactivity.

File: CONTRIBUTING.md
Comment (OriolAbril) [reply CONTRIBUTING.md]: I commented https://github.com/pymc-devs/pymc/pull/5611#discussion_r829896757 trying to not change the meaning. We should first agree on that before working on the wording. 

In case it helps, here is ArviZ guidance: https://arviz-devs.github.io/arviz/contributing/contributing_prs.html

File: CONTRIBUTING.md
Comment (OriolAbril) [reply CONTRIBUTING.md]: see https://github.com/pymc-devs/pymc/pull/5611#discussion_r829891767 first

**Comment by OriolAbril** [comment]
> Why are the tests triggering? I thought they only run if there are changes in .py files?

This https://github.com/pymc-devs/pymc/blob/main/.github/workflows/arviz_compat.yml#L7 paths thing has never worked.

File: CONTRIBUTING.md
Comment (michaelosthege) [reply CONTRIBUTING.md]:
```suggestion
Make sure to mention the issue number in the PR description and commit message.
```

File: CONTRIBUTING.md
Comment (canyon289) [inline CONTRIBUTING.md]:
```suggestion
This will automatically add a link to the PR in the issue ticket as well.
By doing this way others can see that the issue is being worked on and has a pull request.
```

File: CONTRIBUTING.md
Comment (canyon289) [inline CONTRIBUTING.md]: I agree with both comments, the language is harsh and I'm not confident in the proposed process.

Here's my proposal

Any issue without an open pull request is available for work. However if your pull request is more than two weeks old with no activity and someone else expresses interest reserve the right to either close it, or let someone else take it over.

**Comment by canyon289** [comment]
Thanks @michaelosthege for creating this, the good news in all this is theres enough interested people that its actually a problem :) Of course one that still should be solved though, I appreciate you opening this PR.

File: CONTRIBUTING.md
Comment (michaelosthege) [reply CONTRIBUTING.md]:
```suggestion
When you start working working on an issue, open a `Draft` pull request as soon as you're doing the first steps.
Before opening a PR with a new feature, please first open an Issue or Discussion to propose it and discuss the new feature with the maintainers.
```

File: CONTRIBUTING.md
Comment (michaelosthege) [reply CONTRIBUTING.md]: Yeah it is my intention to strongly discourage the reservation behavior.

At the same time, I am disillusioned about lengthy rules and definitions.
Up to today I have not even once opened the contributing guide for ArviZ.

The most important thing is that we agree on the procedure. 2 % of contributors will have read these rules before doing their first PR, or commenting "I want to work on this".

File: CONTRIBUTING.md
Comment (michaelosthege) [inline CONTRIBUTING.md:42]:
```suggestion
1. Read the [Contributing etiquette section](#etiquiette-for-code-contributions).

1. Fork the [project repository](https://github.com/pymc-devs/pymc/) by clicking on the 'Fork' button near the top right of the main repository page. This creates a copy of the code under your GitHub user account.
```

File: CONTRIBUTING.md
Comment (OriolAbril) [reply CONTRIBUTING.md]: Why the duplication? I think I have never used issue numbers in commit messages, should I have been doing that?

Comment (ricardoV94) [reply CONTRIBUTING.md]: You can do that. It does make issue threads awfully verbose if a PR has a lot of rebases + force pushes, because a new automatic message is added every time that is done.

**Review by michaelosthege — COMMENTED** [review]
I see we are not on the same page yet regarding the process.

Before we write up a really long list of rules and defintions that few people are going to read, shall we maybe discuss this in the next lab meeting instead?

File: CONTRIBUTING.md
Comment (OriolAbril) [reply CONTRIBUTING.md]: If I always add the links to all relevant issues in the PR description, is there an advantage in adding them also in commit messages?

File: CONTRIBUTING.md
Comment (OriolAbril) [reply CONTRIBUTING.md]: sounds good

File: CONTRIBUTING.md
Comment (OriolAbril) [reply CONTRIBUTING.md]: > At the same time, I am disillusioned about lengthy rules and definitions.

They should not be excessively long, but they should be explicit and clear. Now that I have some experience in open source I often skip them (if I am fixing a typo in their docs for example) but bad contributing guides have driven me away from more fleshy PRs in some libraries.

> The most important thing is that we agree on the procedure.

200% agreed on this, if we don't agree on this we'll end up giving incoherent advise which can be even worse than having no docs

File: CONTRIBUTING.md
Comment (ricardoV94) [reply CONTRIBUTING.md]: Not really. You can always go from a commit to its PR and then to the original issue

Comment (michaelosthege) [reply CONTRIBUTING.md]: The links from the description can get lost depending on how the PR is merged.
> You can always go from a commit to its PR and then to the original issue

Yes, but not the other way around unless the issue number ends up in a commit message on `main`.
Or at least only with a detour through the PR.

**Comment by michaelosthege** [comment]
I just prepended a commit that fixes the `paths:` filter in our CI pipelines.
The CI runs of this PR are evidence that the change works:
* The edit of the workflow definitions triggered the workflows
* The subsequent commits editing the `CONTRIBUTING.md` only triggered `pre-commit` and docs build.
------
Content-wise, I updated larger sections of the guide (it was really outdated!).

If we don't come to an agreement regarding the "I want to work on this" stuff, I would propose to remove that last commit and merge just the CI fix and reformatting for now.

**Comment by OriolAbril** [comment]
> Content-wise, I updated larger sections of the guide (it was really outdated!).

the actual docs might have been kept a bit more to date, but I doubt that. Mid term this page will be gone and be converted to a link to the documentation: #5425

**Comment by michaelosthege** [comment]
I rephrased and reformatted the etiquette section again, based on @canyon289's suggestion:
```
* Any issue without an open pull request is available for work. However if a pull request has no recent activity it may be closed, or taken over by someone else.
* In accordance with the previous point, please don't make unrealistic "_I want to work on this issue_" pledges.
```
...and also removing some things that were duplicated from the PR steps below.

File: .github/workflows/pre-commit.yml
Comment (OriolAbril) [inline .github/workflows/pre-commit.yml]: these changes seem to make no CI at all run for prs submitted from forks: https://github.com/pymc-devs/pymc/pull/5617

**Comment by canyon289** [comment]
@michaelosthege wanted to say I really appreciate you putting this PR together. Even if I have quetsions about specifics I fully agree some additional guidance is needed on both sides, for contributors and for us so things are smoother. 

Thank you prompting this discussion

**Comment by ricardoV94** [comment]
> Also, can we remove the guideline to prefix a draft PR with [WIP] (mentioned a couple of lines below your changes)? We should just use the draft tool from Github, it's more clean.

Did you manage to do this? I don't see it in the changes

**Comment by michaelosthege** [comment]
> > Also, can we remove the guideline to prefix a draft PR with [WIP] (mentioned a couple of lines below your changes)? We should just use the draft tool from Github, it's more clean.
> 
> Did you manage to do this? I don't see it in the changes

Yes, it was the old line 95 https://github.com/pymc-devs/pymc/pull/5611/files#diff-eca12c0a30e25b4b46522ebf89465a03ba72a03f540796c979137931d8f92055L95

File: .github/PULL_REQUEST_TEMPLATE.md
Comment (OriolAbril) [inline .github/PULL_REQUEST_TEMPLATE.md:8]: why remove this?

File: CONTRIBUTING.md
Comment (OriolAbril) [inline CONTRIBUTING.md]: We should set a time here. What does "recent activity" mean? I have no idea and not knowing if they are supposed to respond within 24h or 3 days to prevent their PR from being closed is very stressful.

**Review by OriolAbril — APPROVED** [review]

File: CONTRIBUTING.md
Comment (michaelosthege) [reply CONTRIBUTING.md]: I didn't want to set a time because it depends a lot on context. If it's a PR to fix a failing CI on `main`, we might take over after a few hours, but if it's a PR adding a new complicated step method or distribution we'll often wait weeks.

It's really a matter of good judgement and I wouldn't add a rule preventing that

File: .github/PULL_REQUEST_TEMPLATE.md
Comment (michaelosthege) [reply .github/PULL_REQUEST_TEMPLATE.md:8]: Because they run automatically in the CI.
Ticking the box or not doesn't make a difference. What counts is `pre-commit` being ❌|✔.

Comment (OriolAbril) [reply .github/PULL_REQUEST_TEMPLATE.md:8]: Maybe we could add a link to the checklist instead? As you mentioned before, 2% of users will have read the contributing guide, but maybe 20% read the pull request template and will benefit from the reference/extra context

File: CONTRIBUTING.md
Comment (OriolAbril) [reply CONTRIBUTING.md]: It doesn't need to be a rule. We know that this will depend on context, and it will also depend on availability, if we are all busy with other things or even other PRs we can leave a PR unchecked for longer independently of context. But many (or most) of the readers of this document/section do not have much/any context so we need to provide that:

* Will they be expected to answer in a matter of hours? No except if very rare exceptions you'll probably never find youself in. There may be PRs where such a fast cadence is needed if they are needed to fix CI on `main` or to fix the documentation website.
* Can they expect to work on a PR only once every 2 weeks? No
* If the PR takes longer than expected and you are required to stop working on it for a while, what happens? Answered already in a point below.

Even expert pymc users might not have any context when it comes to contributing to open source.

Comment (michaelosthege) [reply CONTRIBUTING.md]:
```suggestion
* Any issue without an open pull request is available for work. However if a pull request has no recent activity it may be closed, or taken over by someone else. What is considered "recent" will depend on the urgency, and everybody is asked to remain calm and use good judgement.
```
Is this better?
Otherwise, please apply another suggestion.

File: .github/PULL_REQUEST_TEMPLATE.md
Comment (michaelosthege) [reply .github/PULL_REQUEST_TEMPLATE.md:8]: I added the link above the checklist.

File: CONTRIBUTING.md
Comment (canyon289) [reply CONTRIBUTING.md]: How's this? Also note I'm totally fine if all this is rejected and we go with something else. Just trying to add ideas to consider :D 

```suggestion
* Any issue without an open pull request is available for work. 
   * If a pull request has no recent activity it may be closed, or taken over by someone else.
   * The specific timeframe for recent is hard to define as it depends on the contributor the specific code change, and other contextual factors. As a rule of thumb in a normal pull request with no other blockers there typically is activity every couple of days.
   * The core devs will make their best judgement when opting to close PRs.
```

**Comment by michaelosthege** [comment]
After #5619 was just merged, I rebased this branch.
I squashed the 2nd PR template commit into the first and applied @canyon289's suggestion in the other commit.

Looks like the new workflow filters are doing their job. No tests were triggered by these Markdown-only changes :)

__EDIT:__ Just trailing whitespace again 🙄

File: CONTRIBUTING.md
Comment (canyon289) [inline CONTRIBUTING.md]:
```suggestion
* When you start working working on an issue, open a `Draft` pull request as soon as you make your first commit (see steps below).
```

File: CONTRIBUTING.md
Comment (canyon289) [inline CONTRIBUTING.md]:
```suggestion
* Before opening a PR with a new feature, please make a proposal by opening an [issue](https://github.com/pymc-devs/pymc/issues) or [Discussion](https://github.com/pymc-devs/pymc/discussions) with the maintainers. Depending on the proposal we might direct you to other places such as [`pymc-experimental`](https://github.com/pymc-devs/pymc-experimental) or [`pymc-examples`](https://github.com/pymc-devs/pymc-examples).
```

File: CONTRIBUTING.md
Comment (canyon289) [inline CONTRIBUTING.md]:
```suggestion
   * The core devs will make their best judgement when opting to close PRs or reassign them to others.
```

File: CONTRIBUTING.md
Comment (canyon289) [inline CONTRIBUTING.md]:
```suggestion
* If unsure if an issue ticket is available feel free to ask in the issue ticket. Note however, that per the previous point an open pull request is way to claim an issue ticket. Please do not make unrealistic pledges in the issue tickets.
```

File: CONTRIBUTING.md
Comment (canyon289) [inline CONTRIBUTING.md]:
```suggestion
* It's okay if you are delayed or need to take a break, but please leave a comment in the pull request if you cannot get it to a state where it can be merged. Depending on the change (urgent bugfix vs. new feature) the core devs can determine if the PR needs to be reassigned to get the work done.
```

File: CONTRIBUTING.md
Comment (canyon289) [inline CONTRIBUTING.md]:
```suggestion
1. Read the [Etiquette for code contributions](#etiquette-for-code-contributions).
```

File: CONTRIBUTING.md
Comment (canyon289) [inline CONTRIBUTING.md]:
```suggestion
## Etiquette for code contributions
```

File: CONTRIBUTING.md
Comment (canyon289) [inline CONTRIBUTING.md:122]:
```suggestion
*  Please select "Create draft pull request" in the dropdown menu when opening your pull request to indicate a work in progress. This most importantly will add a claim to the issue ticket if there is one. It also will help avoid duplicated work, to get early input on implementation details or API/functionality, or to seek collaborators.
  *  In the pull request form in Github paste the link to the issue ticket in the body so Github will link the two together
```

**Review by canyon289 — APPROVED** [review]
Approved, i mostly reworded some things for clarity, and added a couple of extra points

Feel free to reject any of my edits and merge

**Pull request merged**
